### PR TITLE
test: static link test binaries

### PR DESCRIPTION
**Pull request by mattklein123**
So they can be used in test environments with different toolchains.

### CODE REVIEW

**Comment by mattklein123** [comment]
@htuch @junr03 @ccaraman

File: bazel/envoy_build_system.bzl [hunk 1]
```diff
@@ -46,11 +55,10 @@ def envoy_linkopts():
             "-lrt",
             "-ldl",
             '-Wl,--hash-style=gnu',
-            "-static-libstdc++",
-            "-static-libgcc",
         ],
-    }) + envoy_select_force_libcpp(["--stdlib=libc++"],
-                                   ["-static-libstdc++", "-static-libgcc"]) \
+    }) + envoy_static_link_libstdcpp_linkopts() \
+    + envoy_select_force_libcpp(["--stdlib=libc++"],
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Do you want to integrate this with envoy_static_link_libstdcpp_linkopts?

**Comment by mattklein123** [comment]
@htuch updated

File: bazel/envoy_build_system.bzl [hunk 2]
```diff
@@ -32,11 +32,21 @@ def envoy_copts(repository, test = False):
...
 
+def envoy_static_link_libstdcpp_linkopts():
+    return select({
+        # OSX provides system and stdc++ libraries dynamically, so they can't be linked statically.
+        "@bazel_tools//tools/osx:darwin": [],
+        "//conditions:default": [
+            "-static-libstdc++",
+            "-static-libgcc",
+        ],
+    }) + envoy_select_force_libcpp(["--stdlib=libc++"],
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Based on the expansion of `envoy_select_force_libcpp`, I think we can drop the prior `select`.

Comment (mattklein123) [reply bazel/envoy_build_system.bzl]: ok let me look

**Comment by mattklein123** [comment]
@htuch updated again

**Pull request merged**